### PR TITLE
Lead

### PR DESCRIPTION
**Pull request by ableksaju**

### CODE REVIEW

File: main/registerAppLifecycleListeners.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import { app } from 'electron';
-import installExtension, { VUEJS3_DEVTOOLS } from 'electron-devtools-installer';
+import installExtension from 'electron-devtools-installer';
```
Comment (akshayitzme) [inline main/registerAppLifecycleListeners.ts]: Shoudnt commit this

File: main/registerAppLifecycleListeners.ts [hunk 2]
```diff
@@ -28,7 +28,7 @@ export default function registerAppLifecycleListeners(main: Main) {
 
 async function installDevTools(main: Main) {
   try {
-    await installExtension(VUEJS3_DEVTOOLS);
+    await installExtension('nhdogjmejiglipccpnnnanhbledajbpd');
```
Comment (akshayitzme) [inline main/registerAppLifecycleListeners.ts]: discard this change

File: models/baseModels/SalesQuote/SalesQuote.ts [hunk 3]
```diff
@@ -48,6 +51,14 @@ export class SalesQuote extends Invoice {
     return invoice;
   }
 
+  static filters: FiltersMap = {};
```
Comment (akshayitzme) [inline models/baseModels/SalesQuote/SalesQuote.ts]: clear this line if `filterMap` isnt used. also remove the import statement

File: package-lock.json
Comment (akshayitzme) [inline package-lock.json:1]: remove this file

File: src/components/Sidebar.vue
Comment (akshayitzme) [inline src/components/Sidebar.vue:1]: always create branch from master and revert this commit

File: src/pages/Dashboard/Dashboard.vue
Comment (akshayitzme) [inline src/pages/Dashboard/Dashboard.vue:1]: revert

File: src/utils/misc.ts
Comment (akshayitzme) [inline src/utils/misc.ts:1]: revert

File: yarn.lock
Comment (akshayitzme) [inline yarn.lock:1]: don't commit this file, delete

File: models/baseModels/Invoice/Invoice.ts [hunk 4]
```diff
@@ -166,29 +166,31 @@ export abstract class Invoice extends Transactional {
     await super.afterSubmit();
 
     // update outstanding amounts
-    await this.fyo.db.update(this.schemaName, {
-      name: this.name as string,
-      outstandingAmount: this.baseGrandTotal!,
-    });
+    if (this.schemaName !== 'SalesQuote') {
```
Comment (akshayitzme) [inline models/baseModels/Invoice/Invoice.ts]: use early exits. also don't hardcode model values like this use `ModelNameEnum` for similar use cases

File: models/baseModels/Lead/Lead.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  validatePhoneNumber,
+} from 'fyo/model/validationFunction';
+import { DocValueMap } from 'fyo/core/types';
+import { ModelNameEnum } from 'models/types';
+
+export class Lead extends Doc {
+  status?: LeadStatus;
+
+  getCustomerDoc(): Lead | undefined {
+    if (!this.name) return;
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts:22]: use braces `{}`

File: models/baseModels/Lead/Lead.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    } as DocValueMap;
+    const newPartyDoc = this.fyo.doc.getNewDoc(
+      ModelNameEnum.Party,
+      partyDocData
+    ) as Lead;
+    return newPartyDoc;
+  }
+
+  getSalesQuoteDoc(itemData: { name: string; rate: number }): Lead | undefined {
+    if (!this.name) return;
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts:42]: use braces `{}`

File: models/baseModels/Lead/Lead.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    const partyDocData = {
+      ...docData,
+      role: 'Customer',
+      fromLead: this.name,
+    } as DocValueMap;
+    const newPartyDoc = this.fyo.doc.getNewDoc(
+      ModelNameEnum.Party,
+      partyDocData
+    ) as Lead;
+    return newPartyDoc;
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts:38]: skip a line after each variable declarations. for better readability

File: models/baseModels/Lead/Lead.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            referenceType: 'Lead',
+          };
+          const lead = fyo.doc.getNewDoc('SalesQuote', data);
+          if (!lead.name) {
+            return;
+          }
+          await router.push(`/edit/SalesQuote/${lead.name}`);
+        },
+      },
+    ];
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts]: move these functions to [models/helpers.ts](https://github.com/AbleKSaju/books/blob/3e3750da1e2470f5dafd7748fb35f480bf2079a3/models/helpers.ts) and import them. 

reference:  https://github.com/AbleKSaju/books/blob/3e3750da1e2470f5dafd7748fb35f480bf2079a3/models/baseModels/SalesQuote/SalesQuote.ts#L75-L78

File: models/baseModels/Lead/Lead.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    };
+  }
+
+  readOnly: ReadOnlyMap = {
+    unit: () => this.inserted,
+    itemType: () => this.inserted,
+    trackItem: () => this.inserted,
+    hasBatch: () => this.inserted,
+    hasSerialNumber: () => this.inserted,
+  };
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts]: unwanted code

File: models/types.ts [hunk 10]
```diff
@@ -1,4 +1,5 @@
 export type InvoiceStatus = 'Draft' | 'Saved' | 'Unpaid' | 'Cancelled' | 'Paid' | 'Return' | 'ReturnIssued';
+export type LeadStatuses = 'Replied' | 'Opportunity' | 'Interested' | 'Converted' | 'Quotation' | 'DonotContact';
```
Comment (akshayitzme) [inline models/types.ts]: this should be called `LeadStatus`

File: models/helpers.ts [hunk 11]
```diff
@@ -229,22 +236,55 @@ export function getTransactionStatusColumn(): ColumnConfig {
...
 export const statusColor: Record<
-  DocStatus | InvoiceStatus,
+  DocStatus | InvoiceStatus | LeadStatus,
   string | undefined
 > = {
   '': 'gray',
   Draft: 'gray',
+  Open: 'gray',
+  Replied: 'yellow',
+  Opportunity: 'yellow',
```
Comment (akshayitzme) [inline models/helpers.ts:305]: why are status colors mentioned here and below? define it only once and at one place only

File: models/helpers.ts [hunk 12]
```diff
@@ -271,6 +311,37 @@ export function getStatusText(status: DocStatus | InvoiceStatus): string {
...
+      return t`Do not Contact`;
+    default:
+      return '';
+  }
+}
+
+export function getLeadStatus(
+  doc?: Lead | Doc | RenderData
+): LeadStatus | DocStatus {
+  if (!doc || !doc.status) {
```
Comment (akshayitzme) [inline models/helpers.ts]: check only if `doc` is `undefined`. can ignore `!doc.status`

File: src/utils/sidebarConfig.ts [hunk 13]
```diff
@@ -258,6 +258,12 @@ function getCompleteSidebar(): SidebarConfig {
           schemaName: 'Party',
           filters: { role: 'Both' },
         },
+        {
+          label: t`Lead`,
+          name: 'lead',
+          route: '/list/Lead',
+          schemaName: 'Lead',
+        },
```
Comment (akshayitzme) [inline src/utils/sidebarConfig.ts]: move `Lead` to `Sales` section. Also set a feature flag for it to enable it on [AccountingSettings.ts](https://github.com/AbleKSaju/books/blob/3e3750da1e2470f5dafd7748fb35f480bf2079a3/models/baseModels/AccountingSettings/AccountingSettings.ts)

File: models/baseModels/Lead/Lead.ts [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    const docData = this.getValidDict(true, true);
+
+    if (!docData) {
+      return;
+    }
+    const partyDocData = {
+      ...docData,
+      role: 'Customer',
+      fromLead: this.name,
+    } as DocValueMap;
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts:33]: Phone number is not passed to `Party` doc when creating via actions

File: models/regionalModels/in/Party.ts [hunk 15]
```diff
@@ -18,5 +19,8 @@ export class Party extends BaseParty {
 
   hidden: HiddenMap = {
     gstin: () => (this.gstType as GSTType) !== 'Registered Regular',
+    fromLead: () => {
+      return this.fromLead ? false : true;
+    },
```
Comment (akshayitzme) [inline models/regionalModels/in/Party.ts]: can be simply refactored to
```ts
fromLead: () => !this.fromLead,
```

File: models/baseModels/Lead/Lead.ts [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    const newPartyDoc = this.fyo.doc.getNewDoc(
+      ModelNameEnum.Party,
+      partyDocData
+    ) as Lead;
+    return newPartyDoc;
+  }
+
+  getSalesQuoteDoc(itemData: { name: string; rate: number }): Lead | undefined {
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts:41]: why mention `itemData` here?.  Ideally `getSalesQuoteDoc` shouldn't be receiving any props.

File: models/baseModels/Lead/Lead.ts [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    const newSalesQuoteDoc = this.fyo.doc.getNewDoc(ModelNameEnum.SalesQuote, {
+      ...docData,
+      party: docData.name,
+      referenceType: ModelNameEnum.Lead,
+      items: [
+        {
+          item: itemData.name,
+          rate: itemData.rate,
+        },
+      ],
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts:58]: unnecessary assignment to `item`

File: models/baseModels/Lead/Lead.ts [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    }) as Lead;
+    return newSalesQuoteDoc;
+  }
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts]: entire function is unnecessary

File: models/baseModels/Lead/Lead.ts [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    return [
+      {
+        group: fyo.t`Create`,
+        label: fyo.t`Customer`,
+        action: async (doc: Doc, router) => {
+          const lead = fyo.doc.getNewDoc('Party', {
+            ...doc.getValidDict(),
+            fromLead: doc.name,
+            role: 'Customer',
+          });
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts]: Phone number isnt passed to `Party` doc when creating new doc

File: models/baseModels/Lead/Lead.ts [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+      ...docData,
+      role: 'Customer',
+      fromLead: this.name,
+    } as DocValueMap;
+    const newPartyDoc = this.fyo.doc.getNewDoc(
+      ModelNameEnum.Party,
+      partyDocData
+    ) as Lead;
+    return newPartyDoc;
+  }
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts]: Unnecessary function

File: models/baseModels/Lead/Lead.ts [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+            party: doc.name,
+            referenceType: 'Lead',
+          };
+          const lead = fyo.doc.getNewDoc('SalesQuote', data);
+          if (!lead.name) {
+            return;
+          }
+          await router.push(`/edit/SalesQuote/${lead.name}`);
+        },
+      },
```
Comment (akshayitzme) [inline models/baseModels/Lead/Lead.ts]: Split these into functions in [helpers.ts](https://github.com/AbleKSaju/books/blob/3e3750da1e2470f5dafd7748fb35f480bf2079a3/models/helpers.ts) and import it here.

File: models/baseModels/Party/Party.ts [hunk 22]
```diff
@@ -125,6 +128,23 @@ export class Party extends Doc {
     };
   }
 
+  async afterDelete() {
+    await super.afterDelete();
+    const leadData = await this.fyo.doc.getDoc(ModelNameEnum.Lead, this.name);
+    await leadData.setAndSync('status', 'Interested');
+  }
```
Comment (akshayitzme) [inline models/baseModels/Party/Party.ts:138]: can be refactored to 

```ts
await this.setAndSync('status', 'Interested');
```

File: models/baseModels/Party/Party.ts [hunk 23]
```diff
@@ -125,6 +128,23 @@ export class Party extends Doc {
...
+  async afterSync() {
+    await super.afterSync();
+
+    if (!this.fromLead) {
+      return;
+    }
+
+    const leadData = await this.fyo.doc.getDoc(ModelNameEnum.Lead, this.name);
+    await leadData.setAndSync('status', 'Converted');
+  }
```
Comment (akshayitzme) [inline models/baseModels/Party/Party.ts:147]: can be refactored to 

```ts
    await this.setAndSync('status', 'Converted');
```

File: models/baseModels/SalesQuote/SalesQuote.ts [hunk 24]
```diff
@@ -48,6 +51,14 @@ export class SalesQuote extends Invoice {
...
+
+  async afterSubmit(): Promise<void> {
+    await super.afterSubmit();
+    const leadData = await this.fyo.doc.getDoc(ModelNameEnum.Lead, this.party);
+    await leadData.setAndSync('status', 'Quotation');
+  }
```
Comment (akshayitzme) [inline models/baseModels/SalesQuote/SalesQuote.ts]: These lines of code should only be performed if `SalesQuote` is created from `Lead`

You can get the linked leadDoc like below,
```ts
await this.loadAndGetLink('party')
```

File: models/baseModels/SalesQuote/SalesQuote.ts [hunk 25]
```diff
@@ -48,6 +51,14 @@ export class SalesQuote extends Invoice {
     return invoice;
   }
 
+  static filters: FiltersMap = {};
+
+  async afterSubmit(): Promise<void> {
+    await super.afterSubmit();
+    const leadData = await this.fyo.doc.getDoc(ModelNameEnum.Lead, this.party);
```
Comment (akshayitzme) [inline models/baseModels/SalesQuote/SalesQuote.ts]: you can get the linked doc by
```ts
this.loadAndGetLink('party')
```

File: schemas/app/Lead.json [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    {
+      "fieldname": "address",
+      "label": "Address",
+      "fieldtype": "Link",
+      "target": "Address",
+      "create": true,
+      "section": "Contacts"
+    }
+  ],
+  "quickEditFields": ["address"],
```
Comment (akshayitzme) [inline schemas/app/Lead.json]: `quickEditFields` is unnecessary here

Comment (ableksaju) [reply schemas/app/Lead.json]: Resolved

File: fyo/model/types.ts [hunk 27]
```diff
@@ -117,3 +117,12 @@ export type DocStatus =
   | 'NotSaved'
   | 'Submitted'
   | 'Cancelled';
+export type LeadStatus =
```
Comment (akshayitzme) [inline fyo/model/types.ts]: Leave a line in between them

File: models/baseModels/Invoice/Invoice.ts [hunk 28]
```diff
@@ -166,6 +166,9 @@ export abstract class Invoice extends Transactional {
     await super.afterSubmit();
 
     // update outstanding amounts
+    if (this.schemaName === 'SalesQuote') {
+      return;
+    }
```
Comment (akshayitzme) [inline models/baseModels/Invoice/Invoice.ts]: Place these lines before the comment to not be confused as its part of the below code, outstanding amount

File: models/baseModels/SalesQuote/SalesQuote.ts [hunk 29]
```diff
@@ -48,6 +55,18 @@ export class SalesQuote extends Invoice {
...
 
+  static filters: FiltersMap = {
+    numberSeries: (doc: Doc) => ({ referenceType: doc.schemaName }),
+  };
+
+  async afterSubmit(): Promise<void> {
+    await super.afterSubmit();
+    if (this.referenceType == ModelNameEnum.Lead) {
+      const leadData = (await this.loadAndGetLink('party')) as Lead;
+      await leadData.setAndSync('status', 'Quotation');
```
Comment (akshayitzme) [inline models/baseModels/SalesQuote/SalesQuote.ts]: variable name should be identical to the doctype. also wrong assertion as `Lead`

File: models/helpers.ts [hunk 30]
```diff
@@ -108,6 +119,43 @@ export function getMakeInvoiceAction(
...
+        fromLead: doc.name,
+        phone: doc.mobile as string,
+        role: 'Customer',
+      });
+      if (!lead.name) {
+        return;
+      }
+      await router.push(`/edit/Party/${lead.name}`);
+    },
+  };
```
Comment (akshayitzme) [inline models/helpers.ts]: variable name should be identical to the doctype.

File: models/helpers.ts [hunk 31]
```diff
@@ -108,6 +119,43 @@ export function getMakeInvoiceAction(
...
+}
+
+export function getSalesQuoteAction(fyo: Fyo): Action {
+  return {
+    group: fyo.t`Create`,
+    label: fyo.t`Sales Quote`,
+    action: async (doc, router) => {
+      const data: { party: string | undefined; referenceType: string } = {
+        party: doc.name,
+        referenceType: 'Lead',
```
Comment (akshayitzme) [inline models/helpers.ts]: use `ModelNameEnum` here

File: models/helpers.ts [hunk 32]
```diff
@@ -108,6 +119,43 @@ export function getMakeInvoiceAction(
...
+      };
+      const lead = fyo.doc.getNewDoc(ModelNameEnum.SalesQuote, data);
+      if (!lead.name) {
+        return;
+      }
+      await router.push(`/edit/SalesQuote/${lead.name}`);
```
Comment (akshayitzme) [inline models/helpers.ts]: variable name should be identical to the doctype.

File: models/baseModels/tests/testLead.spec.ts [hunk 33]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+test('create test docs for Lead', async (t) => {
+  await fyo.doc.getNewDoc(ModelNameEnum.Lead, leadData).sync();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.Lead, leadData.name),
+    `dummy item ${leadData.name} exists`
+  );
+});
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:29]: need to change the name of the test to something suitable. need to change the message that checks documents existence

File: models/baseModels/tests/testLead.spec.ts [hunk 34]
```diff
@@ -0,0 +1,125 @@
...
+
+test('create Customer', async (t) => {
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:31]: name of the test doesn't accurately describe the test case

File: models/baseModels/tests/testLead.spec.ts [hunk 35]
```diff
@@ -0,0 +1,125 @@
...
+});
+
+test('create test docs for Item', async (t) => {
+  await fyo.doc.getNewDoc(ModelNameEnum.Item, itemData).sync();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.Item, itemData.name),
+    `dummy item ${itemData.name}  exists`
+  );
+});
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:66]: test name doesn't accurately describe the actions performed. you can also wrap the creation of test docs in a separate test case in the beginning of the file

File: models/baseModels/tests/testLead.spec.ts [hunk 36]
```diff
@@ -0,0 +1,125 @@
...
+
+test('create SalesQuote', async (t) => {
+  const leadDoc = (await fyo.doc.getDoc(ModelNameEnum.Lead, 'name2')) as Lead;
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:69]: test name does not accurately describe the test case. Here you are creating a Sales Quote from the Lead document. that should be specified in the test name

File: models/baseModels/tests/testLead.spec.ts [hunk 37]
```diff
@@ -0,0 +1,125 @@
...
+        item: itemData.name,
+        rate: itemData.rate,
+      },
+    ],
+  }) as Lead;
+
+  t.equals(
+    leadDoc.status,
+    'Converted',
+    'Before SalesQuote created the status must be Open'
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:87]: ❌  ~Before SalesQuote created the status must be Open~

✅  status must be Open before SQUOT is created

File: models/baseModels/tests/testLead.spec.ts [hunk 38]
```diff
@@ -0,0 +1,125 @@
...
+  );
+  await newSalesQuoteDoc.sync();
+  await newSalesQuoteDoc.submit();
+
+  t.equals(
+    leadDoc.status,
+    'Quotation',
+    'After SalesQuote created the status change to Quotation'
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:95]: ❌ ~After SalesQuote created the status change to Quotation~
✅ status should change to Quotation after SQUOT submission

File: models/baseModels/tests/testLead.spec.ts [hunk 39]
```diff
@@ -0,0 +1,125 @@
...
+  );
+
+  t.ok(
+    await fyo.db.exists(ModelNameEnum.SalesQuote, newSalesQuoteDoc.name),
+    'SalesQuote Created from Lead'
+  );
+});
+
+test('remove Customer', async (t) => {
```
Comment (akshayitzme) [inline models/baseModels/tests/testLead.spec.ts:104]: test name doesn't accurately describe the test case.

File: schemas/app/SalesQuote.json [hunk 40]
```diff
@@ -15,11 +15,30 @@
       "default": "SQUOT-",
       "section": "Default"
     },
+    {
+      "fieldname": "referenceType",
+      "label": "Type",
+      "placeholder": "Type",
+      "fieldtype": "Select",
+      "readOnly": true,
```
Comment (akshayitzme) [inline schemas/app/SalesQuote.json:23]: value should be selectable by the user

**Pull request closed without merging**